Do our own call into Python for DTensor dispatch

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #166370
* #166369
* #166368

Another incremental step: don't deal with custom ops on the critical
path, let the dispatcher do that. Take control of calling into Python
on the critical path as well.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

swolchok committed:
Do our own call into Python for DTensor dispatch

Another incremental step: don't deal with custom ops on the critical
path, let the dispatcher do that. Take control of calling into Python
on the critical path as well.

[ghstack-poisoned]

swolchok committed:
fix bottom-of-stack PR on "Do our own call into Python for DTensor dispatch"


Another incremental step: don't deal with custom ops on the critical
path, let the dispatcher do that. Take control of calling into Python
on the critical path as well.

[ghstack-poisoned]

swolchok committed:
Update on "Do our own call into Python for DTensor dispatch"


Another incremental step: don't deal with custom ops on the critical
path, let the dispatcher do that. Take control of calling into Python
on the critical path as well.

[ghstack-poisoned]

swolchok commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -771,7 +771,10 @@ static Tensor make_tensor_for_subclass_helper(
-      at::DataPtr{nullptr, options.device()},
+      // REVIEW: without the device change here,
+      // test_dtensor_save_load breaks. I'm not at all sure why it was
+      // OK before.
```
based on test failures, changing the device to meta here clearly isn't correct . will go back.

swolchok committed:
put it back on "Do our own call into Python for DTensor dispatch"


Another incremental step: don't deal with custom ops on the critical
path, let the dispatcher do that. Take control of calling into Python
on the critical path as well.

[ghstack-poisoned]

swolchok commented:
running into some weird coupling here. `test_dtensor_save_load` is broken. (repro: `python test/distributed/tensor/test_dtensor.py -k test_dtensor_save_load`, you need to patch the test to reduce world_size to 1 if you have only one GPU)
Apparently, the presence of `DTensor.__torch_dispatch__` somehow causes DTensors' storage to have a valid data_ptr and thus be compatible with serialization. This PR deletes `DTensor.__torch_dispatch__` and breaks it; I know it's the mere presence of `DTensor.__torch_dispatch__` because 1) putting it back fixes the test 2) printing inside `DTensor.__torch_dispatch__` shows that it's not called before or during serialization.

swolchok committed:
partially fix pickling, but tests are still broken on "Do our own call into Python for DTensor dispatch"


Another incremental step: don't deal with custom ops on the critical
path, let the dispatcher do that. Take control of calling into Python
on the critical path as well.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
messed up amending on "Do our own call into Python for DTensor dispatch"


Another incremental step: don't deal with custom ops on the critical
path, let the dispatcher do that. Take control of calling into Python
on the critical path as well.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
set the DTensor dispatch key in make_wrapper_subclass if needed on "Do our own call into Python for DTensor dispatch"


Another incremental step: don't deal with custom ops on the critical
path, let the dispatcher do that. Take control of calling into Python
on the critical path as well.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
rebase, significant fixes to #166370 and following PR on "Do our own call into Python for DTensor dispatch"


Another incremental step: don't deal with custom ops on the critical
path, let the dispatcher do that. Take control of calling into Python
on the critical path as well.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
fix dtensor dispatch key PRs on "Do our own call into Python for DTensor dispatch"


Another incremental step: don't deal with custom ops on the critical
path, let the dispatcher do that. Take control of calling into Python
on the critical path as well.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
Update on "Do our own call into Python for DTensor dispatch"


Another incremental step: don't deal with custom ops on the critical
path, let the dispatcher do that. Take control of calling into Python
on the critical path as well.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
reproduce dispatch key setup from python fallback in dtensor fallback. fix our "upcall" from Python to DTensor dispatch key to be an explicit redispatch so that the preceding fix doesn't break it. on "Do our own call into Python for DTensor dispatch"


Another incremental step: don't deal with custom ops on the critical
path, let the dispatcher do that. Take control of calling into Python
on the critical path as well.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
fixed dependency on "Do our own call into Python for DTensor dispatch"


Another incremental step: don't deal with custom ops on the critical
path, let the dispatcher do that. Take control of calling into Python
on the critical path as well.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok commented:
looks like the test failure is telling us we need to respect monkey-patching of custom_op_handlers. this will cost us a little performance but I guess it's what we have to do

swolchok committed:
respect monkey-patching of custom_op_handlers on "Do our own call into Python for DTensor dispatch"


Another incremental step: don't deal with custom ops on the critical
path, let the dispatcher do that. Take control of calling into Python
on the critical path as well.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
fix dependency on "Do our own call into Python for DTensor dispatch"


Another incremental step: don't deal with custom ops on the critical
path, let the dispatcher do that. Take control of calling into Python
on the critical path as well.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
Update on "Do our own call into Python for DTensor dispatch"


Another incremental step: don't deal with custom ops on the critical
path, let the dispatcher do that. Take control of calling into Python
on the critical path as well.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
fix several bugs in #166371 that tests found on "Do our own call into Python for DTensor dispatch"


Another incremental step: don't deal with custom ops on the critical
path, let the dispatcher do that. Take control of calling into Python
on the critical path as well.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok commented:
noting that basic distributed tests (the two jobs whose names mention `test(distributed, `) are green

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -865,13 +877,19 @@ static PyObject* THPVariable_make_wrapper_subclass(
+  std::optional<DispatchKeySet> extra_dispatch_keys =
+      r.toDispatchKeySetOptional(13);
+  if (is_dtensor) {
+    extra_dispatch_keys = extra_dispatch_keys.value_or(DispatchKeySet())
```
You have both Python and DTensor keys here?
Why not just one?

swolchok commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1010,6 +1094,70 @@ static c10::SymDimVector tuple_to_symintlist(PyObject* obj) {
+// This is much simpler than the __torch_function__ paths precisely
+// because it doesn't have to worry about anything to do with dispatch
+// or __torch_function__.
+void callDTensorCustomOpHandler(
```
hmm, given that we have to respect monkey patching of _custom_op_handlers, I think that having the explicit hardcoded custom op handler path 1) is confusing 2) doesn't respect monkey patching that removes the hardcoded op handlers. I'll remove it.

swolchok commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1010,6 +1094,70 @@ static c10::SymDimVector tuple_to_symintlist(PyObject* obj) {
+         c10::DispatchKey::Python) |
+     c10::DispatchKeySet(c10::DispatchKey::Python));
+
+// This is much simpler than the __torch_function__ paths precisely
```
grepping test/ for NotImplemented, it looks like we probably don't have DTensor-specific NotImplemented tests. will have to add one. Do you have any examples of code you would expect to work (and ideally a pointer to how it works in the `__torch_dispatch__` regime)?

swolchok committed:
remove hardcoded custom op handling. sending for posterity as I plan to experiment with a different approach that doesn't use a DTensor dispatch key on "Do our own call into Python for DTensor dispatch"


Another incremental step: don't deal with custom ops on the critical
path, let the dispatcher do that. Take control of calling into Python
on the critical path as well.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok commented on torch/distributed/tensor/_dispatch.py:
```diff
@@ -135,6 +135,16 @@ def _allow_implicit_replication(self) -> bool:
     def _allow_implicit_replication(self, value: bool) -> None:
         return torch._C._set_dtensor_allow_implicit_replication(value)
 
+    def custom_op_handler(
```
whoops, forgot to remove this

swolchok commented:
superseded by #167051
